model/task/task.go

```diff
@@ -561,6 +562,92 @@ func (t *Task) MarkEnd(finishTime time.Time, detail *apimodels.TaskEndDetail) er
 
 }
 
+func (t *Task) UpdateDisplayTask() error {
+	if !t.DisplayOnly {
+		return fmt.Errorf("%s is not a display task", t.Id)
+	}
+
+	statuses := []string{}
+	catcher := grip.NewBasicCatcher()
+	var timeTaken time.Duration
+	for _, taskId := range t.ExecutionTasks {
+		execTask, err := FindOne(ById(taskId))
+		catcher.Add(err)
+
+		// if any of the execution tasks are scheduled, the display task is too
+		if execTask.Activated {
+			t.Activated = true
+		}
+
+		// the display task's status will be the highest priority of its exec tasks
+		if execTask != nil {
+			statuses = append(statuses, execTask.ResultStatus())
+		}
+
+		// add up the duration of the execution tasks
+		timeTaken += execTask.TimeTaken
+	}
+	if catcher.HasErrors() {
+		return catcher.Resolve()
+	}
+	if len(statuses) > 0 {
+		sort.Sort(byPriority(statuses))
+		t.Status = statuses[0]
+	}
+
+	t.TimeTaken = timeTaken
+
+	return UpdateOne(
+		bson.M{
+			IdKey: t.Id,
+		},
+		bson.M{
+			"$set": bson.M{
+				StatusKey:    t.Status,
+				ActivatedKey: t.Activated,
+				TimeTakenKey: t.TimeTaken,
+			},
+		})
+}
+
+type byPriority []string
```

Comment:
I don't 100% understand why this sorting is required (the UI?

I also worry about the potential increased cost of adding new statuses (or wanting to add user defined statuses in the future.)

Reply:
It's to handle the scenario where execution tasks in a single display task have different result statuses. For example a display task that contains a successful and failed task should display failed in the UI. It does increase the cost of adding new statuses, but it's necessary cost